### PR TITLE
Fix docker build OOMing in CI for arm64 builds

### PR DESCRIPTION
**Pull request by erikjohnston**
Not sure if we just want to default to true?

### CODE REVIEW

File: changelog.d/14173.docker [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix docker build OOMing in CI for amr64 builds.
```
Comment (babolivier) [inline changelog.d/14173.docker:1]:
```suggestion
Fix docker build OOMing in CI for arm64 builds.
```

**Review by babolivier — APPROVED** [review]

File: docker/Dockerfile [hunk 2]
```diff
@@ -108,6 +108,11 @@ RUN mkdir /rust /cargo
 
 RUN curl -sSf https://sh.rustup.rs | sh -s -- -y --no-modify-path --default-toolchain stable
 
+
+# arm64 builds consume a lot of memory if `CARGO_NET_GIT_FETCH_WITH_CLI` is not
+# set to true, so we expose it as a build-arg.
+ARG CARGO_NET_GIT_FETCH_WITH_CLI
```
Comment (DMRobertson) [inline docker/Dockerfile:114]: do ARGs become environment variables for the directives below them? (otherwise how does this work?!)

Comment (erikjohnston) [reply docker/Dockerfile:114]: Yup, I believe so!

Comment (babolivier) [reply docker/Dockerfile:114]: Good point, it doesn't look like it from a quick local test I've done.

Comment (babolivier) [reply docker/Dockerfile:114]: Good point, it doesn't look like it from a quick local test:

```
[18:09:36] babolivier@Noah:~/Documents/test$ cat Dockerfile 
FROM alpine
ARG TEST
COPY ./test.sh /test.sh
ENTRYPOINT ["/test.sh"]
[18:10:28] babolivier@Noah:~/Documents/test$ cat test.sh 
#!/bin/sh
env
[18:10:40] babolivier@Noah:~/Documents/test$ docker build -t test -f Dockerfile --build-arg=TEST=foo .
Sending build context to Docker daemon  3.072kB
Step 1/4 : FROM alpine
 ---> 9c6f07244728
Step 2/4 : ARG TEST
 ---> Using cache
 ---> 69700947c516
Step 3/4 : COPY ./test.sh /test.sh
 ---> Using cache
 ---> 2bb37955a594
Step 4/4 : ENTRYPOINT ["/test.sh"]
 ---> Using cache
 ---> 120f3318464d
Successfully built 120f3318464d
Successfully tagged test:latest
[18:12:11] babolivier@Noah:~/Documents/test$ docker run test --build-arg=TEST=foo
HOSTNAME=fce46e554943
SHLVL=1
HOME=/root
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PWD=/
[18:12:17] babolivier@Noah:~/Documents/test$ 
```

Comment (erikjohnston) [reply docker/Dockerfile:114]: nyargh what

Comment (erikjohnston) [reply docker/Dockerfile:114]: Oh for gods sake, `RUN $TEST` would do the right thing, but we actually need an `ENV TEST=$TEST` I think

**Review by DMRobertson — APPROVED** [review]

**Comment by DMRobertson** [comment]
```
#26 7.217       cargo rustc --lib --message-format=json-render-diagnostics --manifest-path rust/Cargo.toml --release -v --features pyo3/extension-module -- --crate-type cdylib
  #26 7.217           Updating crates.io index
  #26 7.217       error: Unable to update registry `crates-io`
  #26 7.217       
  #26 7.217       Caused by:
  #26 7.217         failed to fetch `[https://github.com/rust-lang/crates.io-index`](https://github.com/rust-lang/crates.io-index%60)
  #26 7.217       
  #26 7.217       Caused by:
  #26 7.217         error in environment variable `CARGO_NET_GIT_FETCH_WITH_CLI`: provided string was not `true` or `false`
```

File: docker/Dockerfile [hunk 3]
```diff
@@ -108,6 +108,12 @@ RUN mkdir /rust /cargo
 
 RUN curl -sSf https://sh.rustup.rs | sh -s -- -y --no-modify-path --default-toolchain stable
 
+
+# arm64 builds consume a lot of memory if `CARGO_NET_GIT_FETCH_WITH_CLI` is not
+# set to true, so we expose it as a build-arg.
+ARG CARGO_NET_GIT_FETCH_WITH_CLI=false
+ENV CARGO_NET_GIT_FETCH_WITH_CLI=$CARGO_NET_GIT_FETCH_WITH_CLI
```
Comment (richvdh) [inline docker/Dockerfile:115]: er, note that `ENV` defines a variable that will be present in containers started from the final image, which I don't think is really what you want here.

(I guess it won't do any harm, but it does seem like an odd variable for us to export to our image).

Instead you have to do `CARGO_NET_GIT_FETCH_WITH_CLI=$CARGO_NET_GIT_FETCH_WITH_CLI` before the `pip` that needs it.

Comment (erikjohnston) [reply docker/Dockerfile:115]: Yeah, though `ENV` should only affect the current build stage and so not be in the final image.

Comment (richvdh) [reply docker/Dockerfile:115]: ah yes, fair

**Pull request merged**